# In-silico Size Selection and Deduplication

## Introduction

Non-Invasive Prenatal Testing (NIPT) is an alternative to amniocentesis, allowing for genotyping of a fetus using a plasma sample from the pregnant mother[^1]. The results of genotyping this sample is an amalgamation of reads from both the fetus and the mother. The mother's DNA predominates, making up approximately 80% of the available reads. The fetal fraction is approximately 20% of available reads, but half of those are from the mother and cannot be identified as fetal. This leaves approximately 10% of the DNA as an identifiable fetal fraction. 

Enhancing the fetal fraction makes fetal genotyping easier, and can be done by using In-silico Size Selection (ISS). The idea is that on average the fetal reads are shorter than maternal reads[^2]. By selecting only the shorter reads the percentage of reads that are fetal increases, allowing for a larger (and easier to genotype) fetal fraction.

This pipeline takes paired end reads generated by an Illumina MiSeq, selects for reads shorter than or equal to the size limit(s) you select, aligns them, and then deduplicates the results.

[^1]: Lo, Y.M., et al., Presence of fetal DNA in maternal plasma and serum. Lancet, 1997. 350(9076): p. 485-7.
[^2]: Chan, K.C., et al., Size distributions of maternal and fetal DNA in maternal plasma. Clin Chem, 2004. 50(1): p. 88-92.

### In-silico Size Selection

In-silico Size Selection (ISS) is the process of selecting for shorter reads in a sample. This is performed via Fastp[^3],[^4] using the `--length_limit` flag, followed by the maximum length of nucleotides you are looking for (inclusive.) Because Fastp is 



### Deduplication

## Requirements

* Python 3.10 or higher running on MacOS, a flavor of Linux, or Windows with a Linux Subsystem.
* The following command line tools installed (all tools were installed via bioconda and use their versioning):

	* Fastp
	* Burrow-Wheeler Aligner MEM (BWA)
	* Genome Analysis ToolKit 4 (GATK4)
	* Gencore

* An appropriate human genome reference library, in our case, hg19 or hg38.
* Fastq.gz files from Paired End samples run on an Illumina MiSeq.

## Input files

### Sample IDs 
Sample IDs may include a dash (`-`), but must not contain an underscore (`_`). If you use an underscore, the sample ID will get truncated. Ex: Sample-ID works, Sample_ID will be truncated to "Sample".

### File Structure

![An image of the file structure necessary to run this script](https://artemisplayground.s3.us-east-2.amazonaws.com/static/share/images/nipt-informatics/parent_file_structure.png)

This is the file structure required for this script to run. The gzip'd fastq files from the Illumina MiSeq go in samples, the script's output files are found in the results folder. If you are using a reference genome other than hg38, replace the hg38 folder with the reference genome of your choice. You will also need to update the script accordingly. (More on that below.) The Python script needs to stay in the parent folder.

### Files

#### Fastq files
The fastq.gz files for the deduplication/size selection process can be found here (where Illumina is the base folder where you store your MiSeq results):  

```
Illumina/Name_of_run/Data/Intensities/BaseCalls/sample-name_S#_L001_R1_001.fastq.gz 
Illumina/Name_of_run/Data/Intensities/BaseCalls/sample-name_S#_L001_R2_001.fastq.gz
```
These files should be placed in the samples folder. No prior change or manipulation of the files is necessary. No subfolders are necessary, though files in subfolders will be found and processed as well. This script only works with paired end reads, so you will need R1 and R2 files for each sample.

![An image of the file structure highlighting the samples folder necessary to run this script](https://artemisplayground.s3.us-east-2.amazonaws.com/static/share/images/nipt-informatics/samples_file_structure.png)

#### Reference Genome

This script has only been tested using hg19 and hg38 


## Output files

The script returns the results in a subfolder of the sample's ID in the results folder. Three files as returned: the size selected results
